janusgraph-core/src/main/java/org/janusgraph/graphdb/transaction/vertexcache/CaffeineVertexCache.java

```diff
@@ -0,0 +1,133 @@
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package org.janusgraph.graphdb.transaction.vertexcache;
+import com.github.benmanes.caffeine.cache.Cache;
+import com.github.benmanes.caffeine.cache.Caffeine;
+import com.github.benmanes.caffeine.cache.RemovalCause;
+import com.github.benmanes.caffeine.cache.RemovalListener;
+import com.github.benmanes.caffeine.cache.stats.CacheStats;
+import com.google.common.base.Preconditions;
+import org.jctools.maps.NonBlockingHashMapLong;
+import org.janusgraph.graphdb.internal.InternalVertex;
+import org.janusgraph.graphdb.vertices.AbstractVertex;
+import org.janusgraph.util.datastructures.Retriever;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import javax.annotation.Nonnull;
+import javax.annotation.Nullable;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.concurrent.ConcurrentMap;
+
+public class CaffeineVertexCache implements VertexCache {
+    private static final Logger log =
+        LoggerFactory.getLogger(CaffeineVertexCache.class);
+
+    private final ConcurrentMap<Long, InternalVertex> volatileVertices;
+    private final Cache<Long, InternalVertex> cache;
+
+    private long createdTime; // the timestamp when the cache was created.
+
+    class CaffeineRemovalListener implements RemovalListener<Long, InternalVertex> {
+
+        @Override
+        public void onRemoval(@Nullable Long key,
+                              @Nullable InternalVertex internalVertex,
+                              @Nonnull RemovalCause removalCause) {
+            if (removalCause == RemovalCause.EXPLICIT) {
+                assert volatileVertices.isEmpty();
+                return;
+            }
+            assert (removalCause == RemovalCause.SIZE || removalCause == RemovalCause.REPLACED) : "Cause: " + removalCause;
+            InternalVertex v = internalVertex;
+            if (((AbstractVertex) v).isTxOpen() && (v.isModified() || v.isRemoved())) {
+                volatileVertices.putIfAbsent(key, v);
+            }
+        }
+    }
+
+    public CaffeineVertexCache(final long maxCacheSize, final int initialDirtySize) {
+        volatileVertices = new NonBlockingHashMapLong<>(initialDirtySize);
+        log.debug("Created dirty vertex map with initial size {}", initialDirtySize);
+
+        cache = Caffeine.newBuilder().maximumSize(maxCacheSize)
+            .removalListener(new CaffeineRemovalListener())
+            .recordStats()
```

Comment:
I see that cache statistics is always enabled here, even so many applications won't use it. It's definitely helpful for analyses but many times it won't be needed. Moreover, I don't know if there is any memory / performance overhead with having statistics enabled. 
What I noticed is that you report statistics only when closing cache and only when debug is enabled. Thus, I would recommend here to enable `.recordStats()` only when debug log level is enabled. I.e. something like this:
```
Caffeine<Long, InternalVertex> cacheBuilder = Caffeine.newBuilder().maximumSize(maxCacheSize)
    .removalListener(new CaffeineRemovalListener())
    .executor(Runnable::run); // according to the https://github.com/ben-manes/caffeine/discussions/757
if(log.isDebugEnabled()){
    cacheBuilder = cacheBuilder.recordStats();
}
cache = cacheBuilder.build();
```
Another option (maybe even better but not necessary to do this in this PR) is to add a new configuration option to JanusGraph to enable / disable cache stats recording.

Reply:
Also, in case you do this with debug enabled / disabled then make sure in `close` method there is no NPE thrown by any chance (I just didn't check if that is even possible). But for additional protection you can wrap that code in additional if.
I.e.:
```
if(log.isDebugEnabled()){
    long end = System.currentTimeMillis();
    CacheStats stats = cache.stats();
    log.debug("Caffeine cache (lifespan: {}ms) stats: {}", end - createdTime, stats);  
}
```

Reply:
I haven't benchmarked this for a while, but have recently been advising a developer writing a similar cache for .NET. He recently ported Java's `LongAdder` to replace an equivalent of `AtomicLong` for statistics, since I mentioned that was a likely bottleneck. You can see how there is a degradation but the penalty is not noticeable in an application as the performance is very high. (https://github.com/bitfaster/BitFaster.Caching/pull/219)